cmd/pop/main.go

```diff
@@ -0,0 +1,428 @@
+		}
+	}
+
+	// After completion, print the downloaded array to stdout as JSON
+	s, err := json.MarshalIndent(&downloaded, "", "    ")
+	if err != nil {
+		return err
+	}
+
+	msgDir := filepath.Dir(args.messagesFile)
+	failedFile := filepath.Join(msgDir, "failed_downloading")
+	failedJSON, err := json.MarshalIndent(&failed, "", "    ")
+	if err != nil {
+		return err
+	}
+
+	// This is not a failure condition, since the failed ones may be retried
+	_ = os.WriteFile(failedFile, failedJSON, 0o600)
+
+	fmt.Println(string(s))
+	return nil
+}
+
+func runUpload(args uploadArgs) error {
+	if args.messagesFile == "" {
+		return fmt.Errorf("you must provide a messages file")
+	}
+
+	if args.signedDir == "" {
+		return fmt.Errorf("you must provide a directory for the signed packages")
+	}
+
+	ctx := context.Background()
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return err
+	}
+
+	msgBytes, err := os.ReadFile(args.messagesFile)
+	if err != nil {
+		return err
+	}
+
+	envelopes := []Envelope{}
+	failed := []Envelope{}
+	successful := []Envelope{}
+	var errs error
+
+	if err := json.Unmarshal(msgBytes, &envelopes); err != nil {
+		return err
+	}
+
+	fail := func(e error, f ...Envelope) {
+		for _, c := range f {
+			failed = append(failed, c)
+		}
+		errs = errors.Join(errs, err)
+	}
+
+	nameToEnvelopes := make(map[string][]Envelope)
+	for _, envelope := range envelopes {
+		message, err := envelope.GetMessageContent()
+		if err != nil {
+			fail(err, envelope)
+			continue
+		}
+
+		nameToEnvelopes[message.Artifact.Name] = append(nameToEnvelopes[message.Artifact.Name], envelope)
+	}
+
+	client, err := azblob.NewClient(prodAccountName, cred, nil)
+	if err != nil {
+		return err
+	}
+
+	for pkgBasename, envelopes := range nameToEnvelopes {
+		signedPkgFilename := filepath.Join(args.signedDir, pkgBasename)
+		if _, err := os.Stat(signedPkgFilename); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("filename not found in the signing directory; signing likely failed for '%s'", pkgBasename))
+			continue
+		}
+
+		envelope, err := resolveDuplicates(envelopes)
+		if err != nil {
+			fail(err, envelopes...)
+			continue
+		}
+
+		message, err := envelope.GetMessageContent()
+		if err != nil {
+			fail(err, envelopes...)
+			continue
+		}
+
+		pkg := message.Spec.Pkg
+		version := fmt.Sprintf("%s+azure", message.Spec.Tag)
+		distro := message.Spec.Distro
+		pkgOS := message.Spec.OS()
+		sanitizedArch := strings.ReplaceAll(message.Spec.Arch, "/", "")
+
+		storagePath := fmt.Sprintf("%s/%s/%s/%s_%s/%s", pkg, version, distro, pkgOS, sanitizedArch, pkgBasename)
+		fmt.Fprintln(os.Stderr, storagePath)
+
+		f, err := os.Open(signedPkgFilename)
+		if err != nil {
+			fail(err, envelope)
+			continue
+		}
+
+		if _, err := client.UploadFile(ctx, prodContainerName, storagePath, f, &azblob.UploadFileOptions{
+			Metadata: map[string]*string{sha256Key: &message.Artifact.Sha256Sum},
```

Comment:
Was the digest recalculated after signing?

Reply:
Fixed, thank you!